### PR TITLE
feat: oauth for builder

### PR DESCRIPTION
**Pull request by rswanson**
### TL;DR

Added OAuth2 authentication for the builder's communication with Quincey.

### What changed?

- Added OAuth2 dependencies to the builder crate
- Introduced new OAuth-related configuration parameters
- Implemented OAuth token fetching in the submit task
- Updated the Quincey API call to include the OAuth bearer token

### How to test?

1. Set the new OAuth environment variables:
   - OAUTH_CLIENT_ID
   - OAUTH_CLIENT_SECRET
   - OAUTH_AUTHENTICATE_URL
   - OAUTH_TOKEN_URL
2. Ensure the Quincey service is configured to accept OAuth tokens
3. Run the builder and verify it can successfully authenticate and communicate with Quincey

### Why make this change?

This change enhances the security of the communication between the builder and Quincey by implementing OAuth2 authentication. It ensures that only authorized builders can interact with Quincey, improving the overall security posture of the system.

### CODE REVIEW

File: crates/builder/src/tasks/submit.rs [hunk 1]
```diff
@@ -11,6 +11,10 @@ use tokio::{sync::mpsc, task::JoinHandle};
 use tracing::{debug, error, instrument, trace};
 use zenith_types::{SignRequest, SignResponse, Zenith};
 
+use oauth2::basic::BasicClient;
+use oauth2::reqwest::http_client;
+use oauth2::{AuthUrl, ClientId, ClientSecret, StandardTokenResponse, TokenResponse, TokenUrl};
```
Comment (prestwich) [inline crates/builder/src/tasks/submit.rs:16]: make into 1 use statement for the crate

remove wrapping newlines

File: crates/builder/src/tasks/submit.rs [hunk 2]
```diff
@@ -59,6 +66,26 @@ impl SubmitTask {
...
+        StandardTokenResponse<oauth2::EmptyExtraTokenFields, oauth2::basic::BasicTokenType>,
+    > {
+        let client = BasicClient::new(
+            ClientId::new(self.config.oauth_client_id.clone()),
+            Some(ClientSecret::new(self.config.oauth_client_secret.clone())),
+            AuthUrl::new(self.config.oauth_authenticate_url.clone())?,
+            Some(TokenUrl::new(self.config.oauth_token_url.clone())?),
+        );
+
+        let token_result: StandardTokenResponse<
```
Comment (prestwich) [inline crates/builder/src/tasks/submit.rs:81]: does this type need to be explicit?

File: crates/builder/src/tasks/submit.rs [hunk 3]
```diff
@@ -59,6 +66,26 @@ impl SubmitTask {
         serde_json::from_slice(&body).map_err(Into::into)
     }
 
+    async fn fetch_oauth_token(
+        &self,
+    ) -> eyre::Result<
+        StandardTokenResponse<oauth2::EmptyExtraTokenFields, oauth2::basic::BasicTokenType>,
```
Comment (prestwich) [inline crates/builder/src/tasks/submit.rs:72]: replace qualified path with use statement for the inner types

**Review by prestwich — CHANGES_REQUESTED** [review]
style

File: crates/builder/src/tasks/submit.rs [hunk 4]
```diff
@@ -43,10 +47,13 @@ impl SubmitTask {
             "pinging quincey for signature"
         );
 
+        let token = self.fetch_oauth_token().await?;
+
         let resp: reqwest::Response = self
             .client
             .post(self.config.quincey_url.as_ref())
             .json(sig_request)
+            .bearer_auth(token.access_token().secret())
```
Comment (anna-carroll) [inline crates/builder/src/tasks/submit.rs:56]: should we also address tx pool oath in the same PR?

we don't necessarily need to worry about efficiency of streamlining to one oath call for both tx pool & quincey within the example builder but would be a nice-to-have if it were simple

File: crates/builder/src/tasks/submit.rs (hunk 3 above)
Comment (rswanson) [reply crates/builder/src/tasks/submit.rs:72]: i am unsure i know what this means

File: crates/builder/src/tasks/submit.rs (hunk 4 above)
Comment (rswanson) [reply crates/builder/src/tasks/submit.rs:56]: currently we don't call the protected endpoints (`/get-order(s)`) so there isn't anything to update. Adding the functionality itself would be outside the scope of this

File: crates/builder/src/tasks/submit.rs [hunk 5]
```diff
@@ -59,6 +65,24 @@ impl SubmitTask {
...
+        let client = BasicClient::new(
+            ClientId::new(self.config.oauth_client_id.clone()),
+            Some(ClientSecret::new(self.config.oauth_client_secret.clone())),
+            AuthUrl::new(self.config.oauth_authenticate_url.clone())?,
+            Some(TokenUrl::new(self.config.oauth_token_url.clone())?),
+        );
+
+        let token_result = client
+            .exchange_client_credentials()
+            .add_extra_param("audience", "https://transactions.holesky.signet.sh")
```
Comment (prestwich) [inline crates/builder/src/tasks/submit.rs:80]: should this be configurable? if not, should they be module-level constants?

**Review by prestwich — APPROVED** [review]

**Review by Evalir — COMMENTED** [review]
seems tests are failing?

**Pull request merged**